modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java

```diff
@@ -274,33 +294,59 @@ public CompletableFuture<long[]> executeBatchAsync(@Nullable Transaction transac
         touchAndCloseIfExpired();
 
         if (!busyLock.enterBusy()) {
-            return CompletableFuture.failedFuture(new SqlException(SESSION_CLOSED_ERR, "Session is closed."));
+            return CompletableFuture.failedFuture(sessionIsClosedException());
         }
 
         try {
-            QueryContext ctx = QueryContext.create(Set.of(SqlQueryType.DML), transaction);
+            SqlProperties properties = SqlPropertiesHelper.newBuilder()
+                    .set(QueryProperty.ALLOWED_QUERY_TYPES, EnumSet.of(SqlQueryType.DML))
+                    .build();
 
             var counters = new LongArrayList(batch.size());
-            CompletableFuture<Void> tail = CompletableFuture.completedFuture(null);
-            ArrayList<CompletableFuture<Void>> batchFuts = new ArrayList<>(batch.size());
+            CompletableFuture<?> tail = CompletableFuture.completedFuture(null);
+            ArrayList<CompletableFuture<?>> batchFuts = new ArrayList<>(batch.size());
 
             for (int i = 0; i < batch.size(); ++i) {
                 Object[] args = batch.get(i).toArray();
 
-                final var qryFut = tail
-                        .thenCompose(v -> qryProc.querySingleAsync(sessionId, ctx, transactions, query, args));
-
-                tail = qryFut.thenCompose(cur -> cur.requestNextAsync(1))
-                        .thenAccept(page -> {
-                            validateDmlResult(page);
-
-                            counters.add((long) page.items().get(0).get(0));
-                        })
-                        .whenComplete((v, ex) -> {
-                            if (ExceptionUtils.unwrapCause(ex) instanceof CancellationException) {
-                                qryFut.cancel(false);
-                            }
-                        });
+                tail = tail.thenCompose(v -> {
+                    if (!busyLock.enterBusy()) {
+                        return CompletableFuture.failedFuture(sessionIsClosedException());
+                    }
+
+                    try {
+                        return qryProc.querySingleAsync(properties, transactions, (InternalTransaction) transaction, query, args)
+                                .thenCompose(cursor -> {
+                                    if (!busyLock.enterBusy()) {
+                                        return CompletableFuture.failedFuture(sessionIsClosedException());
```

Comment:
Should we close cursor here? or if cursor.requestNextAsync(1) failed due to any reason?

Reply:
It seems the same applies to `executeAsync`.

Reply:
totally agree, should check it myself ( 

fixed